<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>con-gen-csu - 16&nbsp; Estimating Site Frequency Spectra and a few applications thereof</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../nmfs-bioinf/mega-post.html" rel="next">
<link href="../nmfs-bioinf/handling-vcf-files.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../nmfs-bioinf/sfs-fst.html"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Estimating Site Frequency Spectra and a few applications thereof</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../nmfs-bioinf/quarto-static/fisheries_header_logo_jul2019.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="../">con-gen-csu</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/eriqande/con-gen-csu" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/open-on-demand-alpine.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">OpenOnDemand on Alpine</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/quick-unix-review.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Quick Unix Review</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/shell-prog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Shell Programming</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/awk-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">A Brief <code>awk</code> Intro</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/scripts-and-functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Bash scripts and functions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/slurm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Sedna and SLURM intro</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/sbatch.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Submitting jobs with <code>sbatch</code></span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/slurm-arrays.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Slurm Job Arrays</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/sequence-alignment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Sequence alignment and handling BAM files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/bioinf-formats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Bioinformatic file formats</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/snake.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Snakemake Narrative and Topics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/snakemake-relevant-python.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Snakemake-relevant Python for R Users</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/snakemake-embellishments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Important Snakemake Embellishments</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/variant-calling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Variant calling</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/handling-vcf-files.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Basic Handling of VCF files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/sfs-fst.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Estimating Site Frequency Spectra and a few applications thereof</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../nmfs-bioinf/mega-post.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Exploring and Using the <code>mega-post-bcf-exploratory-snakeflows</code></span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#what-is-the-sfs" id="toc-what-is-the-sfs" class="nav-link active" data-scroll-target="#what-is-the-sfs"><span class="header-section-number">16.1</span> What is the SFS?</a></li>
  <li><a href="#why-is-it-useful" id="toc-why-is-it-useful" class="nav-link" data-scroll-target="#why-is-it-useful"><span class="header-section-number">16.2</span> Why is it useful?</a></li>
  <li><a href="#more-on-the-sfs-itself-with-simple-examples" id="toc-more-on-the-sfs-itself-with-simple-examples" class="nav-link" data-scroll-target="#more-on-the-sfs-itself-with-simple-examples"><span class="header-section-number">16.3</span> More on the SFS itself with simple examples</a></li>
  <li><a href="#uncertainty-in-the-genotypes" id="toc-uncertainty-in-the-genotypes" class="nav-link" data-scroll-target="#uncertainty-in-the-genotypes"><span class="header-section-number">16.4</span> Uncertainty in the genotypes</a></li>
  <li><a href="#angsd-dosaf" id="toc-angsd-dosaf" class="nav-link" data-scroll-target="#angsd-dosaf"><span class="header-section-number">16.5</span> ANGSD doSaf</a></li>
  <li><a href="#running-angsd-dosaf" id="toc-running-angsd-dosaf" class="nav-link" data-scroll-target="#running-angsd-dosaf"><span class="header-section-number">16.6</span> Running ANGSD doSaf</a>
  <ul class="collapse">
  <li><a href="#cloning-the-workflow" id="toc-cloning-the-workflow" class="nav-link" data-scroll-target="#cloning-the-workflow"><span class="header-section-number">16.6.1</span> Cloning the workflow</a></li>
  <li><a href="#symlinking-the-results-in-our-con-gen-csu-results" id="toc-symlinking-the-results-in-our-con-gen-csu-results" class="nav-link" data-scroll-target="#symlinking-the-results-in-our-con-gen-csu-results"><span class="header-section-number">16.6.2</span> Symlinking the results in our con-gen-csu results</a></li>
  <li><a href="#the-config-for-our-trinity-river-chinook" id="toc-the-config-for-our-trinity-river-chinook" class="nav-link" data-scroll-target="#the-config-for-our-trinity-river-chinook"><span class="header-section-number">16.6.3</span> The config for our Trinity River Chinook</a></li>
  <li><a href="#a-simple-dry-run" id="toc-a-simple-dry-run" class="nav-link" data-scroll-target="#a-simple-dry-run"><span class="header-section-number">16.6.4</span> A Simple Dry run</a></li>
  <li><a href="#only-run-things-as-far-as-the-calc_saf-rule" id="toc-only-run-things-as-far-as-the-calc_saf-rule" class="nav-link" data-scroll-target="#only-run-things-as-far-as-the-calc_saf-rule"><span class="header-section-number">16.6.5</span> Only run things as far as the <code>calc_saf</code> rule</a></li>
  <li><a href="#make-a-picture-of-part-of-that-file" id="toc-make-a-picture-of-part-of-that-file" class="nav-link" data-scroll-target="#make-a-picture-of-part-of-that-file"><span class="header-section-number">16.6.6</span> Make a picture of part of that file</a></li>
  </ul></li>
  <li><a href="#installing-winsfs" id="toc-installing-winsfs" class="nav-link" data-scroll-target="#installing-winsfs"><span class="header-section-number">16.7</span> Installing winsfs</a></li>
  <li><a href="#back-to-lcwgs-pw-fst" id="toc-back-to-lcwgs-pw-fst" class="nav-link" data-scroll-target="#back-to-lcwgs-pw-fst"><span class="header-section-number">16.8</span> Back to lcwgs-pw-fst</a></li>
  <li><a href="#calculating-some-1-d-sfss" id="toc-calculating-some-1-d-sfss" class="nav-link" data-scroll-target="#calculating-some-1-d-sfss"><span class="header-section-number">16.9</span> Calculating some 1-D SFSs</a>
  <ul class="collapse">
  <li><a href="#lets-do-it-by-hand" id="toc-lets-do-it-by-hand" class="nav-link" data-scroll-target="#lets-do-it-by-hand"><span class="header-section-number">16.9.1</span> Let’s do it by hand</a></li>
  <li><a href="#lets-calculate-all-the-1-d-sfs-using-the-snakemake-workflow" id="toc-lets-calculate-all-the-1-d-sfs-using-the-snakemake-workflow" class="nav-link" data-scroll-target="#lets-calculate-all-the-1-d-sfs-using-the-snakemake-workflow"><span class="header-section-number">16.9.2</span> Let’s calculate all the 1-D SFS using the snakemake workflow</a></li>
  </ul></li>
  <li><a href="#running-the-rest-of-the-fst-workflow" id="toc-running-the-rest-of-the-fst-workflow" class="nav-link" data-scroll-target="#running-the-rest-of-the-fst-workflow"><span class="header-section-number">16.10</span> Running the rest of the Fst workflow</a>
  <ul class="collapse">
  <li><a href="#plotting-the-sliding-windows" id="toc-plotting-the-sliding-windows" class="nav-link" data-scroll-target="#plotting-the-sliding-windows"><span class="header-section-number">16.10.1</span> Plotting the Sliding Windows</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/eriqande/con-gen-csu/edit/main/nmfs-bioinf/sfs-fst.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/eriqande/con-gen-csu/issues/new/choose" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sfs-fst" class="quarto-section-identifier"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Estimating Site Frequency Spectra and a few applications thereof</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="what-is-the-sfs" class="level2" data-number="16.1">
<h2 data-number="16.1" class="anchored" data-anchor-id="what-is-the-sfs"><span class="header-section-number">16.1</span> What is the SFS?</h2>
<ul>
<li>Typically the fraction of sites in a sample of <span class="math inline">\(n\)</span> diploids that carry <span class="math inline">\(1, \ldots, 2n-1\)</span> copies of the derived (as opposed to the ancestral) allele.</li>
<li>With whole genome sequencing data, we can also include the categories of 0 and <span class="math inline">\(2n\)</span> copies of the derived alleles.</li>
</ul>
</section>
<section id="why-is-it-useful" class="level2" data-number="16.2">
<h2 data-number="16.2" class="anchored" data-anchor-id="why-is-it-useful"><span class="header-section-number">16.2</span> Why is it useful?</h2>
<ul>
<li>A lot of population genetic theory has been done about what the SFS should look like under different demographic scenarios and/or selection.</li>
<li>Many population genetic summary statistics can be shown to be functions of the SFS. A worthy read about how nearly all the myriad simple tests for non-neutrality (Tajima’s <span class="math inline">\(D\)</span>, Fay and Wu’s <span class="math inline">\(H\)</span>, etc.) can be written as functions of the SFS is <a href="https://academic.oup.com/genetics/article/183/1/249/6063201">Achaz 2009</a>.</li>
<li>Programs exist (<span class="math inline">\(\partial a \partial i\)</span>, moments, etc.) to use SFS (and particularly multi-dimensional—i.e.&nbsp;multi-population—SFS) to estimate demographic history of species.</li>
<li>Pairswise <span class="math inline">\(F_\mathrm{ST}\)</span> between populations can be computed as a function of the 2-D SFS.</li>
<li>The SFS provides an exceptional amount of data reduction: from terabytes of sequencing data to a handful of numbers. Obviously this discards a lot of information, but for some inferences, the SFS is sufficient or nearly so.</li>
</ul>
</section>
<section id="more-on-the-sfs-itself-with-simple-examples" class="level2" data-number="16.3">
<h2 data-number="16.3" class="anchored" data-anchor-id="more-on-the-sfs-itself-with-simple-examples"><span class="header-section-number">16.3</span> More on the SFS itself with simple examples</h2>
<ul>
<li>Illustrate on the whiteboard with a simple data set assuming fully resolved genotypes.</li>
<li>Discuss the <span class="math inline">\(\theta/k\)</span> result from the neutral coalescent.<br>
</li>
<li>Show how easy it is to estimate SFS with <em>complete</em> and <em>certain</em> genotype data.</li>
<li>At any site, show that the calculation involves lining the genotypes up horizontally, with the cumulative numbers of gene copies in each, stacked vertically above them.</li>
</ul>
</section>
<section id="uncertainty-in-the-genotypes" class="level2" data-number="16.4">
<h2 data-number="16.4" class="anchored" data-anchor-id="uncertainty-in-the-genotypes"><span class="header-section-number">16.4</span> Uncertainty in the genotypes</h2>
<ul>
<li>With low-coverage data we don’t get to observe the genotypes with certainty.</li>
<li>We don’t even know ahead of time if there is a SNP there.</li>
<li>So, the ANGSD approach is to first calculate the likelihood of each number of derived alleles, from <span class="math inline">\(0\)</span> to <span class="math inline">\(2n\)</span>, by considering all the possible underlying genotypes. Storing those results, and then using those to do inference.<br>
</li>
<li>Let’s show what that looks like at a single site on the whiteboard.</li>
</ul>
</section>
<section id="angsd-dosaf" class="level2" data-number="16.5">
<h2 data-number="16.5" class="anchored" data-anchor-id="angsd-dosaf"><span class="header-section-number">16.5</span> ANGSD doSaf</h2>
<ul>
<li><p>One of the options to ANGSD is <code>-doSaf</code>. This creates a “site allele frequency” file, which is a binary file that holds the likelihoods for <em>each site</em> of the number of copies of the derived allele are at that site.</p></li>
<li><p>Documentation for it is at: https://www.popgen.dk/angsd/index.php/SFS_Estimation</p></li>
<li><p>There is also “onboard” documentation with the program itself. So, let us get ANGSD in a conda environment so we can call it. If you don’t already have ANGSD in a conda env, do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do this on a compute node or acompile</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> angsd bioconda::angsd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
<li><p>Once that is done, activate the angsd environment and call the program with the <code>-doSaf</code> option and nothing else to get the onboard help:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># activate the environment</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-n</span> angsd bioconda::angsd</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># call the program with no other args</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">angsd</span> <span class="at">-doSaf</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
</ul>
</section>
<section id="running-angsd-dosaf" class="level2" data-number="16.6">
<h2 data-number="16.6" class="anchored" data-anchor-id="running-angsd-dosaf"><span class="header-section-number">16.6</span> Running ANGSD doSaf</h2>
<p>We are going to run doSaf on our course data (the 16 Chinook salmon from the Trinity River).</p>
<p>As always, before we start doing anything you will want to sync the main branch of your fork of the <code>con-gen-csu</code> repo and then pull changes down to the main branch of the local clone on your cluster.</p>
<p>We are going to run <code>angsd -doSaf</code> within a <a href="https://github.com/eriqande/mega-lcwgs-pw-fst-snakeflow">Snakemake workflow that I have on GitHub</a>. The purpose of the workflow is to compute pairwise <span class="math inline">\(F_\mathrm{ST}\)</span> between different groups of samples, and doing so requires the <code>-doSaf</code> calculation from ANGSD.</p>
<section id="cloning-the-workflow" class="level3" data-number="16.6.1">
<h3 data-number="16.6.1" class="anchored" data-anchor-id="cloning-the-workflow"><span class="header-section-number">16.6.1</span> Cloning the workflow</h3>
<p>So, the first thing that you will need to do is clone the Snakemake workflow. You can fork it (if you like) and then clone your own flow, OR you could just clone the repo as is. Make sure that you DO NOT clone it into your <code>con-gen-csu</code> directory. Rather, it would be better to clone it into your <code>projects</code> or <code>scratch</code> directory.</p>
<p>Cloning <code>eriqande/mega-lcwgs-pw-fst-snakeflow</code> directly, as opposed to forking it and cloning your fork, would look like this:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first cd into your projects and scratch directories</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone git@github.com:eriqande/mega-lcwgs-pw-fst-snakeflow.git</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="symlinking-the-results-in-our-con-gen-csu-results" class="level3" data-number="16.6.2">
<h3 data-number="16.6.2" class="anchored" data-anchor-id="symlinking-the-results-in-our-con-gen-csu-results"><span class="header-section-number">16.6.2</span> Symlinking the results in our con-gen-csu results</h3>
<p>For calculating Fst we will use the BAMs that were created when you ran exercise 008. To make this relatively easier to describe (basically so that the same paths will work for all of us regardless of where everyone’s BAM files are) we will use symbolic links, making a symbolic link to your <code>con-gen-csu</code> directory in your home directory, named <code>CGC</code>.</p>
<p>To do this,</p>
<ol type="1">
<li><p>navigate inside of your <code>con-gen-csu</code> directory,</p></li>
<li><p>do <code>pwd</code>.<br>
</p></li>
<li><p>Copy the resulting absolute path</p></li>
<li><p><code>cd</code> to your <code>mega-lcwgs-pw-fst-snakeflow</code> directory</p></li>
<li><p>Then, in that <code>mega-lcwgs-pw-fst-snakeflow</code> directory, do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> absolute-path-to-con-gen-csu  CGC</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>but replace <code>absolute-path-to-con-gen-csu</code> with the actual path to your <code>con-gen-csu</code> directory.</p></li>
</ol>
<p>For example, when I do this on alpine it looks like:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># here I get the absolute path to con-gen-csu and I copy it</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">%</span> pwd</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/eriq@colostate.edu/projects/con-gen-csu</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># then I change directories to the mega-lcwgs-pw-fst-snakeflow directory</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="ex">%</span> cd /home/eriq@colostate.edu/projects/mega-lcwgs-pw-fst-snakeflow/</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># then I paste the con-gen-csu path in to do:</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /home/eriq@colostate.edu/projects/con-gen-csu CGC</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="the-config-for-our-trinity-river-chinook" class="level3" data-number="16.6.3">
<h3 data-number="16.6.3" class="anchored" data-anchor-id="the-config-for-our-trinity-river-chinook"><span class="header-section-number">16.6.3</span> The config for our Trinity River Chinook</h3>
<p>We have a config file that will let you use your bams through the <code>CGC</code> alias you just made. The config files to allow the <code>mega-lcwgs-pw-fst-snakeflow</code> to use the bams from the Trinity River Chinook salmon are in the directory <code>extras/lcwgs-pw-fst-config</code> in the <code>con-gen-csu</code> repo, which means that from within the <code>mega-lcwgs-pw-fst-snakeflow</code> directory, you can access it via the relative path <code>CGC/extras/lcwgs-pw-fst-config</code>. The config files within that directory, with links to them on GitHub, are:</p>
<ul>
<li><a href="https://github.com/eriqande/con-gen-csu/blob/main/extras/lcwgs-pw-fst-config/config.yaml"><code>config.yaml</code></a>: the main config file for Snakemake to use.</li>
<li><a href="https://github.com/eriqande/con-gen-csu/blob/main/extras/lcwgs-pw-fst-config/bams.tsv"><code>bams.tsv</code></a>: paths to the BAM files for the 16 Chinook salmon, and group designations (spring or fall)</li>
<li><a href="https://github.com/eriqande/con-gen-csu/blob/main/extras/lcwgs-pw-fst-config/chroms.tsv"><code>chroms.tsv</code></a>: summary information about the chromosomes.</li>
<li><a href="https://github.com/eriqande/con-gen-csu/blob/main/extras/lcwgs-pw-fst-config/pwcomps.tsv"><code>pwcomps.tsv</code></a>: description of groups to make pairwise <span class="math inline">\(F_\mathrm{ST}\)</span> comparisons of.</li>
</ul>
</section>
<section id="a-simple-dry-run" class="level3" data-number="16.6.4">
<h3 data-number="16.6.4" class="anchored" data-anchor-id="a-simple-dry-run"><span class="header-section-number">16.6.4</span> A Simple Dry run</h3>
<p>We can do a simple dry run to see which jobs will be run if we were to do all the steps for calculating <span class="math inline">\(F_\mathrm{ST}\)</span>, including <span class="math inline">\(F_\mathrm{ST}\)</span> values in sliding windows.</p>
<p>Do the following after getting onto a compute node with four cores, by, for example, doing <code>acompile -n 4</code> or <code>srun -p atesting -c 4 --pty /bin/bash</code></p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Type this in the mega-lcwgs-pw-fst-snakeflow directory</strong></pre>
</div>
<div class="sourceCode" id="cb6" data-filename="Type this in the mega-lcwgs-pw-fst-snakeflow directory"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate snakemake-8.5.3</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-np</span>  <span class="at">--configfile</span> CGC/extras/lcwgs-pw-fst-config/config.yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Phew! That spits out a lot of different jobs. Let’s quickly have a look at what that is doing by looking at the rulegraph:</p>
<p><img src="./figs/lcwgs-pw-rulegraph.svg" class="img-fluid"></p>
<p>Aha! Most of the jobs are happening after the <code>calc_saf</code> rule, which is what we are focused on.</p>
</section>
<section id="only-run-things-as-far-as-the-calc_saf-rule" class="level3" data-number="16.6.5">
<h3 data-number="16.6.5" class="anchored" data-anchor-id="only-run-things-as-far-as-the-calc_saf-rule"><span class="header-section-number">16.6.5</span> Only run things as far as the <code>calc_saf</code> rule</h3>
<p>This is a good time to talk about a useful option to Snakemake. If you want to only run a workflow up to a certain point, you can provide the <code>--until</code> option, giving it a rule name. So, try:</p>
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>Type this in</strong></pre>
</div>
<div class="sourceCode" id="cb7" data-filename="Type this in"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-np</span> <span class="at">--until</span> calc_saf  <span class="at">--configfile</span> CGC/extras/lcwgs-pw-fst-config/config.yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>That just runs 8 jobs making bamlists (short little jobs) and 8 jobs making SAF files—4 chromosomes in each of two groups.</p>
<p>Let’s see about running all of those:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-p</span> <span class="at">--use-conda</span> <span class="at">--until</span> calc_saf <span class="at">--cores</span> 4 <span class="at">--configfile</span> CGC/extras/lcwgs-pw-fst-config/config.yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>That does not take too long.</p>
<p>Let’s check out the size of the resulting files:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span> <span class="at">--du</span> <span class="at">-h</span> results/BY_CHROM</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Whoa! Those are some pretty big files. For <code>NC_037122.1f5t9</code> for example, I get 56 M and 64 M for spring-run and fall-run respectively.</p>
<p>Let’s do some thinking about that—<code>NC_037122.1f5t9</code> is 4 megabases and each group of fish has 8 diploids, so there are <span class="math inline">\(2 \times 8 + 1 = 17\)</span> possible sample allele frequencies at each base, and each frequency requires 4 bytes to store in memory. Thus we would expect an uncompressed size of this file in megabytes to be:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fl">4e6</span> <span class="sc">*</span> <span class="dv">17</span> <span class="sc">*</span> <span class="dv">4</span> <span class="sc">/</span> <span class="fl">1e6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 272</code></pre>
</div>
</div>
<p>But it is only around 60 Mb, and that is because it is compressed, and the compressed size is about 22% of the uncompressed.</p>
</section>
<section id="make-a-picture-of-part-of-that-file" class="level3" data-number="16.6.6">
<h3 data-number="16.6.6" class="anchored" data-anchor-id="make-a-picture-of-part-of-that-file"><span class="header-section-number">16.6.6</span> Make a picture of part of that file</h3>
<p>In order to get a sense of what this file looks like, let’s extract a piece of it and then plot it.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we can get the first 5000 sites from the saf file using the ANGSD tool</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># realSFS</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="ex">realSFS</span> print results/BY_CHROM/NC_037122.1f5t9/saf/Fall.saf.idx <span class="at">-r</span> NC_037122.1f5t9:1-5000 <span class="op">&gt;</span> fall-first-5k.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now, we have that file in the con-gen-csu repo, so we can have a look at it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the file</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>saf <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"extras/inputs/fall-first-5k.tsv"</span>, <span class="at">col_names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(saf) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chrom"</span>, <span class="st">"pos"</span>, <span class="dv">0</span><span class="sc">:</span><span class="dv">16</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># get a version that is not on the log scale</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>exp_saf <span class="ot">&lt;-</span> saf <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="at">.cols =</span> <span class="st">`</span><span class="at">0</span><span class="st">`</span><span class="sc">:</span><span class="st">`</span><span class="at">16</span><span class="st">`</span>, <span class="at">.fns =</span> exp))</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co"># try plotting the natural scale one</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>saf_long <span class="ot">&lt;-</span> exp_saf <span class="sc">%&gt;%</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span><span class="fu">c</span>(chrom, pos), <span class="at">names_to =</span> <span class="st">"num_derived"</span>, <span class="at">values_to =</span> <span class="st">"likelihood"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">num_derived =</span> <span class="fu">as.integer</span>(num_derived))</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(saf_long) <span class="sc">+</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">x =</span> num_derived, <span class="at">y =</span> pos, <span class="at">fill =</span> likelihood)) <span class="sc">+</span> </span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> <span class="fu">as.integer</span>(pos <span class="sc">/</span> <span class="dv">500</span>), <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="sfs-fst_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>Oh Wow! That is pretty cool! Each facet is 500 bases of the chromosome. The x-axis is number of “derived” alleles in the sample, and the fill color is likelihood. The white bands are omitted bases (likely becuase there are no reads there? or some other feature of the genome there?).</p>
<p>You can download a higher resolution version of that figure from <a href="https://github.com/eriqande/con-gen-csu/blob/main/nmfs-bioinf/figs/saf_file_raster.pdf">here</a>.</p>
<p>These SAF files form the basis for a lot of further inference, which we address next.</p>
</section>
</section>
<section id="installing-winsfs" class="level2" data-number="16.7">
<h2 data-number="16.7" class="anchored" data-anchor-id="installing-winsfs"><span class="header-section-number">16.7</span> Installing winsfs</h2>
<p><code>winsfs</code> (which, by the way, stands for “Windowed Site Frequency Spectrum”) is written in the Rust language. It is not available on conda, but it is relatively easy to build. The default instructions to build it are available at <a href="https://github.com/malthesr/winsfs?tab=readme-ov-file#installation">https://github.com/malthesr/winsfs?tab=readme-ov-file#installation</a>. If you are not on ALPINE, use those directions.</p>
<p>If you are on ALPINE, however, you want the build chain and other stuff to go in your projects directory, not your home directory (which has very little space in it.)</p>
<p>So, if you are on ALPINE you want to follow these steps: first, from a fresh shell and in your home directory, define the environment variables CARGO_HOME and RUSTUP_HOME to be the absolute path to your projects directory. (Otherwise, the rust build chain will eat up all the space in your scrawny home directories).</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do this on acompile</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="ex">acompile</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">CARGO_HOME</span><span class="op">=</span>/projects/eriq@colostate.edu</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">RUSTUP_HOME</span><span class="op">=</span>/projects/eriq@colostate.edu</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then we want to put the build chain there, so we do:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">--proto</span> <span class="st">'=https'</span> <span class="at">--tlsv1.2</span> <span class="at">-sSf</span> https://sh.rustup.rs <span class="kw">|</span> <span class="fu">sh</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>That should tell you that it will install everything into your projects directory. Hit ENTER to agree to continue installing. (It looks like you should choose 1, but you actually want to just hit ENTER…)</p>
<p>That take a minute or two, but should finish, eventually.</p>
<p>Now, we need to set up the environment</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> <span class="va">$CARGO_HOME</span>/env</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co"># and then just to be on the safe side, you should do:</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/.bashrc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>That should be it for getting the build chain, so now you should be able to get winsfs:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">cargo</span> install winsfs-cli</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>When that is done, do <code>which winsfs</code> to confirm where it was built and placed.</p>
<p>Then, do <code>winsfs --help</code> to get the help info.</p>
</section>
<section id="back-to-lcwgs-pw-fst" class="level2" data-number="16.8">
<h2 data-number="16.8" class="anchored" data-anchor-id="back-to-lcwgs-pw-fst"><span class="header-section-number">16.8</span> Back to lcwgs-pw-fst</h2>
<p>If you go back to the shell where you were working on the mega-lcwgs-pw-fst-snakeflow workflow, you will want to <code>source ~/.bashrc</code> to get the path to winsfs, and then you will want to get back onto atesting for working further on this.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">-p</span> atesting <span class="at">-c</span> 4  <span class="at">--pty</span> /bin/bash</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># after you get into that, check to make sure winsfs is available</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="ex">winsfs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="calculating-some-1-d-sfss" class="level2" data-number="16.9">
<h2 data-number="16.9" class="anchored" data-anchor-id="calculating-some-1-d-sfss"><span class="header-section-number">16.9</span> Calculating some 1-D SFSs</h2>
<p>Both <code>realSFS</code> from angsd and <code>winsfs</code> operate similarly for this, but give somewhat different results. If you pass either of them an <code>saf.idx</code> file, they will do the optimization to return a maximum likelihood estimate of the SFS from all the sites in the file.</p>
<section id="lets-do-it-by-hand" class="level3" data-number="16.9.1">
<h3 data-number="16.9.1" class="anchored" data-anchor-id="lets-do-it-by-hand"><span class="header-section-number">16.9.1</span> Let’s do it by hand</h3>
<div class="sourceCode" id="cb19"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># activate your angsd env</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate angsd</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># learn about realSFS options</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="ex">realSFS</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co"># run realSFS on one of the SAF files we made previously</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="ex">realSFS</span> <span class="at">-cores</span> 4  results/BY_CHROM/NC_037122.1f5t9/saf/Fall.saf.idx</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>That takes a bit of time, so we can talk about what it is doing. Note that most of the stuff being written to the screen is going to stderr. The part that goes to stdout is just the line:</p>
<pre><code>3737572.635705 1561.211911 927.350151 465.278375 438.794388 56.110069 390.037928 89.693553 308.188647 21.118673 0.011756 254.166472 79.952746 1.729683 85.879451 134.411897 296.428594</code></pre>
<p>Now, for comparison, let’s do that same thing with <code>winsfs</code>:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">winsfs</span> <span class="at">-t</span> 4  results/BY_CHROM/NC_037122.1f5t9/saf/Fall.saf.idx</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>That happens so fast that you don’t even have time to talk about it!</p>
<p>The output is:</p>
<pre><code>#SHAPE=&lt;17&gt;
3737470.683161 1625.494654 932.293365 540.814038 412.352881 142.438646 287.406187 128.243713 226.198641 65.924898 65.140087 153.462124 110.387933 45.092476 49.428736 148.862580 278.775880</code></pre>
<p>Which is pretty much the same, except it has the added feature of telling us how many values are in the SFS.</p>
<p>And, reactivate your snakemake environment</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate snakemake-8.5.3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="lets-calculate-all-the-1-d-sfs-using-the-snakemake-workflow" class="level3" data-number="16.9.2">
<h3 data-number="16.9.2" class="anchored" data-anchor-id="lets-calculate-all-the-1-d-sfs-using-the-snakemake-workflow"><span class="header-section-number">16.9.2</span> Let’s calculate all the 1-D SFS using the snakemake workflow</h3>
<p>You will need to get the latest update of the mega-lcwgs-pw-fst-snakeflow workflow by syncing your fork (if you have forked it) and then pulling into main.</p>
<p>I have set up two new rules to calculate 1-D SFS in the workflow—not becuase they are needed for calculating Fst, but just for our own edification.</p>
<p>You can run the workflow and get those outputs using:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check with a dry run:</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-np</span> <span class="at">--use-conda</span>  <span class="at">--cores</span> 8 dest_edify_1d <span class="at">--configfile</span> CGC/extras/lcwgs-pw-fst-config/config.yaml</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># if that looks good, do it with:</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-p</span> <span class="at">--use-conda</span>  <span class="at">--cores</span> 8 dest_edify_1d <span class="at">--configfile</span> CGC/extras/lcwgs-pw-fst-config/config.yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>I have copied the results to the course repo so that we can plot them all:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">dir</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="fu">c</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extras/lcwgs-pw-fst-results/one_d_realSFS"</span>,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extras/lcwgs-pw-fst-results/one_d_winsfs"</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.names =</span> <span class="cn">TRUE</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(files) <span class="ot">&lt;-</span> files</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>big_tibble <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  files, </span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) {</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>    lines <span class="ot">&lt;-</span> <span class="fu">read_lines</span>(x)</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>    counts <span class="ot">=</span> <span class="fu">as.numeric</span>(<span class="fu">str_split</span>(lines[<span class="fu">length</span>(lines)], <span class="at">pattern =</span> <span class="st">" +"</span>)[[<span class="dv">1</span>]])</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>    counts <span class="ot">&lt;-</span> counts[<span class="sc">!</span><span class="fu">is.na</span>(counts)]</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">num_derived =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">16</span>,</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">count =</span> counts</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>(<span class="at">.id =</span> <span class="st">"file"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">extract</span>(</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>    file,</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"method"</span>, <span class="st">"chrom"</span>, <span class="st">"ecotype"</span>),</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">regex =</span> <span class="st">"^.*one_d_([a-zA-Z]+)/(.+)---(.+)</span><span class="sc">\\</span><span class="st">.ml"</span></span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the 0 category, cuz it be huge, then plot</span></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>big_tibble <span class="sc">%&gt;%</span></span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(num_derived <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> num_derived, <span class="at">y =</span> count, <span class="at">fill =</span> ecotype)) <span class="sc">+</span></span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> </span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(chrom <span class="sc">~</span> ecotype <span class="sc">+</span> method)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="sfs-fst_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>We see that the results from <code>winsfs</code> are less choppy. This is what the authors of the Rasmussen et al.&nbsp;paper meant when they said that the <code>winsfs</code> approach is less prone to overfitting.</p>
<p>Remember that these are unfolded SFSes where we used the reference genome as the “ancestral” genome.</p>
</section>
</section>
<section id="running-the-rest-of-the-fst-workflow" class="level2" data-number="16.10">
<h2 data-number="16.10" class="anchored" data-anchor-id="running-the-rest-of-the-fst-workflow"><span class="header-section-number">16.10</span> Running the rest of the Fst workflow</h2>
<p>We can let this thing rip with:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode sh code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check it with a dry run</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-np</span> <span class="at">--use-conda</span> <span class="at">--cores</span> 8 <span class="at">--configfile</span> CGC/extras/lcwgs-pw-fst-config/config.yaml</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Let 'er rip</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-p</span> <span class="at">--cores</span> 8 <span class="at">--use-conda</span> <span class="at">--configfile</span> CGC/extras/lcwgs-pw-fst-config/config.yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>While this is running we can look over the steps of the workflow. Basically, they are:</p>
<ol type="1">
<li>Calculate the MLE of the 2-D SFS for each pair of populations (just Fall and Spring here). This is done in <code>rule calc_2dsfs_winsfs</code>.</li>
<li>Fold the resulting 2-D sfs to reflect that we don’t know the ancestral state. Done in <code>rule fold_winsfs</code></li>
<li>Use that MLE of the 2-D SFS as a prior to calculate the posterior number of derived alleles in each population <em>at each base pair</em> and from that, calculate a summary value related to Fst for each base pair. (Done using realSFS in <code>rule calc_fst_binaries_winsfs</code>)</li>
<li>From those summaries at each site, extract a single Fst value for each chromosome. (Done in <code>rule extract_fst_values_winsfs</code>)</li>
<li>From those summaries at each site, calculate Fst in a sliding window for each chromosome. (Done in <code>rule sliding_window_fst_winsfs</code>)</li>
<li>Summarize all that output into just a couple of files, (Done in <code>rule summarise_average_fst_values</code> and <code>rule summarise_sliding_window_fst_values</code>)</li>
</ol>
<section id="plotting-the-sliding-windows" class="level3" data-number="16.10.1">
<h3 data-number="16.10.1" class="anchored" data-anchor-id="plotting-the-sliding-windows"><span class="header-section-number">16.10.1</span> Plotting the Sliding Windows</h3>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>windows <span class="ot">&lt;-</span> <span class="fu">read_table</span>(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"extras/lcwgs-pw-fst-results/summarized/sliding_window_fst/Fall--x--Spring--size-30000--step-5000.tsv"</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">col_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">skip =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">str_detect</span>(X1, <span class="st">"region"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">chrom =</span> X2,</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">window_mid =</span> X3,</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">Fst =</span> X5</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
── Column specification ────────────────────────────────────────────────────────
cols(
  X1 = col_character(),
  X2 = col_character(),
  X3 = col_double(),
  X4 = col_double(),
  X5 = col_double()
)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(windows, <span class="fu">aes</span>(<span class="at">x =</span> window_mid, <span class="at">y =</span> Fst)) <span class="sc">+</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> chrom)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="sfs-fst_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">dir</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="fu">c</span>(</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extras/lcwgs-pw-fst-results/NC_037122.1f5t9/winsfs"</span>,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extras/lcwgs-pw-fst-results/NC_037123.1f10t14/winsfs"</span>,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extras/lcwgs-pw-fst-results/NC_037124.1f8t16/winsfs"</span>,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"extras/lcwgs-pw-fst-results/NC_037125.1f20t24/winsfs"</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">full.names =</span> <span class="cn">TRUE</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(files) <span class="ot">&lt;-</span> files</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>big_tibble <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>  files, </span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) {</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>    lines <span class="ot">&lt;-</span> <span class="fu">read_lines</span>(x)</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>    counts <span class="ot">=</span> <span class="fu">as.numeric</span>(<span class="fu">str_split</span>(lines[<span class="fu">length</span>(lines)], <span class="at">pattern =</span> <span class="st">" +"</span>)[[<span class="dv">1</span>]])</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>    counts <span class="ot">&lt;-</span> counts[<span class="sc">!</span><span class="fu">is.na</span>(counts)]</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">num_derived_Fall =</span> <span class="fu">rep</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">16</span>, <span class="dv">17</span>),</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">num_derived_Spring =</span> <span class="fu">rep</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">16</span>, <span class="at">each =</span> <span class="dv">17</span>),</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">count =</span> counts</span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>(<span class="at">.id =</span> <span class="st">"file"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">extract</span>(</span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>    file,</span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"chrom"</span>),</span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">regex =</span> <span class="st">"^.*(NC_[^/]+)/.+$"</span></span>
<span id="cb30-29"><a href="#cb30-29" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-30"><a href="#cb30-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-31"><a href="#cb30-31" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the 0 category, cuz it be huge, then plot</span></span>
<span id="cb30-32"><a href="#cb30-32" aria-hidden="true" tabindex="-1"></a>big_tibble <span class="sc">%&gt;%</span></span>
<span id="cb30-33"><a href="#cb30-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(num_derived_Fall <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> num_derived_Spring <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb30-34"><a href="#cb30-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> num_derived_Fall, <span class="at">y =</span> num_derived_Spring, <span class="at">fill =</span> count)) <span class="sc">+</span></span>
<span id="cb30-35"><a href="#cb30-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span> </span>
<span id="cb30-36"><a href="#cb30-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> chrom) <span class="sc">+</span></span>
<span id="cb30-37"><a href="#cb30-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">trans=</span><span class="st">"log10"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="sfs-fst_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../nmfs-bioinf/handling-vcf-files.html" class="pagination-link" aria-label="Basic Handling of VCF files">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Basic Handling of VCF files</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../nmfs-bioinf/mega-post.html" class="pagination-link" aria-label="Exploring and Using the `mega-post-bcf-exploratory-snakeflows`">
        <span class="nav-page-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Exploring and Using the <code>mega-post-bcf-exploratory-snakeflows</code></span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/eriqande/con-gen-csu/edit/main/nmfs-bioinf/sfs-fst.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/eriqande/con-gen-csu/issues/new/choose" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>